lib/TPP/Dialect/Xsmm/XsmmUtils.cpp

```diff
@@ -564,6 +570,144 @@ FailureOr<SmallVector<Attribute>> getBrgemmFlags(PatternRewriter &rewriter,
+  llvm::set_intersect(batches, b);
+  llvm::set_intersect(batches, c);
+
+  // A & B red are the reduction dimensions.
+  llvm::SmallDenseSet<int64_t> ra =
+      findIndexingOperand(indexingMaps[0], iterators, red);
+  llvm::SmallDenseSet<int64_t> rb =
+      findIndexingOperand(indexingMaps[1], iterators, red);
+  llvm::set_intersect(ra, rb);
+
+  // Return each set in sorted order.
+  ContractionDimensions dimensions{
+      SmallVector<unsigned, 2>(batches.begin(), batches.end()),
+      SmallVector<unsigned, 2>(ac.begin(), ac.end()),
+      SmallVector<unsigned, 2>(bc.begin(), bc.end()),
+      SmallVector<unsigned, 2>(ra.begin(), ra.end())};
+  llvm::sort(dimensions.batch.begin(), dimensions.batch.end());
+  llvm::sort(dimensions.m.begin(), dimensions.m.end());
+  llvm::sort(dimensions.n.begin(), dimensions.n.end());
+  llvm::sort(dimensions.k.begin(), dimensions.k.end());
+  return dimensions;
+}
+
+std::optional<unsigned> getAffineBinaryOpExprIndex(AffineMap map, int index,
+                                                   MLIRContext *context) {
+  for (unsigned i = 0; i < map.getNumResults(); i++) {
+    auto result = map.getResult(i);
+    if (isa<AffineBinaryOpExpr>(result) &&
+        dyn_cast<AffineBinaryOpExpr>(result).getLHS() ==
+            getAffineDimExpr(index, context)) {
+      return i;
+    }
+  }
+  llvm_unreachable("invalid binary op index");
+}
+
+LogicalResult checkVNNIGemmStructure(PatternRewriter &rewriter,
+                                     linalg::GenericOp linalgOp) {
+  if (linalgOp->getNumOperands() != 3)
+    return failure();
+
+  if (xsmm::utils::getDataType(rewriter, linalgOp.getOperand(0).getType()) !=
+      xsmm::DataTypeAttr::get(rewriter.getContext(), xsmm::DataType::BF16)) {
+    return failure();
+  }
+  auto iteratorTypes = linalgOp.getIteratorTypesArray();
+  if (iteratorTypes.size() < 4)
+    return failure();
+
+  auto contractionDims = inferContractionDims(linalgOp);
+  if (failed(contractionDims))
+    return failure();
+
+  unsigned m = contractionDims->m.back();
+  unsigned n = contractionDims->n.back();
+
+  // m and n dimensions must be parallel dimensions
+  if (!linalg::isParallelIterator(iteratorTypes[m]) ||
+      !linalg::isParallelIterator(iteratorTypes[n])) {
+    return failure();
+  }
+
+  // innermost dimension must be a reduction dimension for VNNI type operations
+  if (!linalg::isReductionIterator(iteratorTypes[iteratorTypes.size() - 1])) {
+    return failure();
+  }
+
+  // get the index of the iterator corresponding to the floordiv operation
+  auto k = contractionDims->k.size() > 0 ? contractionDims->k.back() : 0;
+  auto map1 = linalgOp.getIndexingMapsArray()[1];
+  auto index = getAffineBinaryOpExprIndex(map1, k, linalgOp.getContext());
+  if (!index)
+    return failure();
+
+  // Ensure that the body of the generic operation is mul-add chain
+  // clang-format off
+  using namespace mlir::structured_match;
+  auto hasRightOpChain =
```

Comment:
Nit: I'd call it `isMatmulChain`, because "right" is subjective and non descriptive.